### ISSUE TITLE
Not able to fold set_encoding(constant) -> constant in data tiling path

### ISSUE THREAD

**Issue by hanhanW** [openxla/iree#11360]
Coming from https://github.com/iree-org/iree/pull/11284, the set_encoding(constant) is not folded into a constant. This happens automatically in legacy mmt4d path because the set_encoding op is expand_reshape + transpose. The fusion on tensors can fold `expand_reshape(constant)` into a constant; a pattern transposes the constant.

If we can fold set_encoding(constant) -> constant, we can save ~360 flow.dispatch calls for FP32 mobilebert. There are ~2700 kernel launches in total, which means that we can save ~13% of kernel launch.

**Comment by hanhanW** [comment]
I have interesting prototype that kick in const-eval stuff; we can evaluate how it helps data-tiling: https://github.com/openxla/iree/commit/9769860777bcd24abc1d57cc02be66283e0e92d1 I moved the SetEncoding pass before element-wise fusion and remove one assertion about encoding on HAL types. The total number of elements is the same, so I think it's probably fine?

```cpp
llvm::Optional<int32_t> getEncodingTypeValue(Attribute attr) {
  // TODO(#6762): encoding attribute handling/mapping to enums.
  // assert(!attr && "encoding types other than default not yet supported");
  // Default to IREE_HAL_ENCODING_TYPE_DENSE_ROW_MAJOR for now.
  return 1;
}
```

This is not a right way plumbing things through. It does not work for cross compilation.

Here is what I got for MobileBertFP model (on x86):

- The size of output VMFB files are all 94 MB
- The number of unique dispatches is cut from 56 to 52.
- The number of `flow.dispatch @.*` is cut from 2391 to 2030

This means that we can kill 4 dispatch and reduce 361 kernel launches w/ const eval.

The bad side is that it generates around 360 private util.global variables.

IR dump w/o the commit: https://gist.githubusercontent.com/hanhanW/ba0ae5f6434f28addd2d8f519cfa46a1/raw

IR dump w/ the commit and enabling const-eval: https://gist.githubusercontent.com/hanhanW/d77a150a76ef72e5b7ed7af0eb3c9fd4/raw

 cc @benvanik and @MaheshRavishankar who might be interested in this.

**Comment by benvanik** [comment]
Nice! For the encoding you could use `IREE_HAL_ENCODING_TYPE_OPAQUE` for now (0) when the attribute is set - that ensures that if we cross the boundary into user code (return from a function/etc) the buffer view has the right encoding on it - though we should also probably ensure that we don't encode result tensors (may happen naturally today, not sure).

The number of globals won't matter as we pack them all together later on, so that's fine.

The bigger thing is the non-portable effects this has. We still need to break down the workflow into stages such that we can do the useful transformations early (transpose of entire tensor) while leaving the exact block swizzle patterns to runtime when desired (default). The eval lifting should be able to move the swizzling stuff to initialization time even if we can't fold it.

**Comment by hanhanW** [comment]
Here are some breakdown about packing on constants

Model | Packing on Constants | % of execution time
| -- | -- | -- |
DeepLabV3 | < 1 ms | 3.7%
PoseNetFP32 | 0.65 ms | 2.9 %
MobileSSD | 3.26 ms | 9.8 %
MobileNetV1 | 5.04 ms |  18.3 %
MobileNetV2 | 3.77 ms | 29.4 %
MobileNetV3Small | 0.67 ms | 15.2 %
MobileBert FP32 | 15.42 ms | 8.3 %
BertLargeBatch1 | 304 ms | 0.9 %
BertLargeBatch32 | 600 ms | 0.8 %

**Comment by hanhanW** [comment]
Copy contents from a dup issue (#14317) and track things in this issue.

Hanhan:

The splat constant is a simple case, which can be upstreamed as a folding pattern. For non-splat constants, we can either emulate it. TPP has an [implementation](https://github.com/plaidml/tpp-mlir/blob/26d9e255bdd40b2c5588ab44bffa07f1a384b105/lib/TPP/ConstantFoldPack.cpp#L50); we can upstream it and add the pass to IREE. We don't add it to a folding pattern because that could execute for a long time, which increases compilation time.

Ben:

Note that if this becomes load bearing for performance we'll want to ensure that we are doing some hierarchical packing (macro/micro blocks) - if the exact innermost packing is not statically available then we can't fold it into constants but could still do some outermost packing (into 64x64 blocks, etc). A VMFB generated for one single specific architecture will be fine with basic packing, but a VMFB generated for multiple architectures (including the same major architecture but different features) will need this (think if the packing params differ between little and big ARM cores but the user compiling can't specify which cores they are running on in the iree-compile command line).

**Comment by MaheshRavishankar** [comment]
Some one can clarify these details but I think there are two things we need here

1) We need the constant expression hoisting to encompass the `set_encoding` operations as well so that they are marked for execution.
2) After that the dispatches formed after hoisting are compiled and run on the VMVX backend. This has a "minor" issue. The following is the kind of dispatch that this will result in. Below is the IR after `CPUMaterializeEncoding` pass when compiling for VMVX backend.

```
hal.executable.variant public @vmvx_bytecode_fb, target = <"vmvx", "vmvx-bytecode-fb", {ukernels = true}> {
  hal.executable.export public @simple_matmul_dispatch_0_set_encoding_MATMUL_F32F32F32_LHS_DxD ordinal(0) layout(#hal.pipeline.layout<push_constants = 4, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) {
  ^bb0(%arg0: !hal.device, %arg1: index, %arg2: index, %arg3: index, %arg4: index):
    %x, %y, %z = flow.dispatch.workgroup_count_from_slice %arg1, %arg2, %arg3, %arg4
    hal.return %x, %y, %z : index, index, index
  }
  builtin.module {
    func.func @simple_matmul_dispatch_0_set_encoding_MATMUL_F32F32F32_LHS_DxD() {
       ....
      %padded = tensor.pad %14 low[0, 0] high[%16, %15] {
      ^bb0(%arg0: index, %arg1: index):
        tensor.yield %cst : f32
      } : tensor<5x10xf32> to tensor<?x?xf32>
      %17 = iree_linalg_ext.set_encoding %padded : tensor<?x?xf32> -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<5x10xf32>>>
      flow.dispatch.tensor.store %17, %11, offsets = [0, 0], sizes = [%9, %10], strides = [1, 1] : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<5x10xf32>>> -> !flow.dispatch.tensor<writeonly:tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F3\
2F32F32, role = LHS, original_type = tensor<5x10xf32>>>>{%9, %10}
      return
    }
  }
}
```

This pass interprets the encoding and uses the the `target` attribute on the `hal.executable.variant` to get the target information. Here the target information is related to VMVX backend. Instead we need a way to record that this dispatch needs to use a "target" that is the actual device being targeted (say `cascadelake`). The fix in `CPUMaterializeEncodingPass` is to have a place where the presence of a `target` attribute overrides what is specified in the `hal.executable.variant`.

3) The other remaining question is how to get the const eval to set this attribute at Flow level and propagate it to the HAL for the `CPUMaterializeEncoding` to pick up.

(2) and (3) are related. (1) is fairly independent and can be done after (2) and (3) are figured out.

@benvanik and @stellaraccident . Please verify that this matches what you expect and maybe provide suggestions for how and where to override the `target` attribute.

cc @bjacob

**Comment by bjacob** [comment]
> The fix in CPUMaterializeEncodingPass is to have a place where the presence of a target attribute overrides what is specified in the hal.executable.variant

This should be working like this already: we should always use the HALExecutableTargetAttr::lookup helper, which should check for a target attribute first before walking up to the enclosing variant.

**Comment by MaheshRavishankar** [comment]
> > The fix in CPUMaterializeEncodingPass is to have a place where the presence of a target attribute overrides what is specified in the hal.executable.variant
> 
> This should be working like this already: we should always use the HALExecutableTargetAttr::lookup helper, which should check for a target attribute first before walking up to the enclosing variant.

Ok, then there might be a path here. We can borrow from what was done w.r.t `ukernels` support on CUDA [here](https://github.com/openxla/iree/blob/d37ccadaabf78d80e790e5b111a80a3c0cfff15e/samples/custom_dispatch/cuda/kernels/ukernel_example.mlir#L18) . If the dispatch is being hoisted for const evaluation we can set the target on the `flow.dispatch.region` operation that is formed by extending the `iree_codegen.export_config` to also contain an optional target information.  Somewhere in HAL we can then move that attribute on the surrounding `func.func` or `builtin.module`. Then the encoding will see the specified `target` attribute.

**Comment by benvanik** [comment]
I think what you're reaching for is an override outside of the target specification - whatever is looking for the information you need should be pulling from the most local attribute possible (maybe even on the op you care about, or it's function scope) and only ever sourcing from the target attr as a fallback. Adding an `iree_codegen.encoding_blah` attr in places would let you assign this independently from the target and that's required for const-evaling on a target backend different than the one you are evaluating the code for. The target on the hal.executable.variant will always need to be the real information for the target evaluating the code and you can't change any individual subattribute inside it as it'll lead to all sorts of weird problems (some op/pass may be using the same kind of lookup in order to generate correct code for the evaluating target and not work if there are random changes in the configuration).

^ just saw Mahesh's reply: yeah, I think you could just scope the iree_codegen.export_config and use that instead of replicating the HAL target properties - just always have your code look up the export config first. Maybe export config is a misnomer though as these are really just like LLVM function attributes and we want to be able compile executables that support multiple different final configurations at the same time. "this function has any ops using encoding use these parameters while this other function uses these other parameters"

**Comment by bjacob** [comment]
EDIT - @MaheshRavishankar helpfully points out that there's no constant in my testcase :embarrassed:

I tried a bit, but it's not easy to get the constant expressions analysis to pick this up. The initial hurdle is that `HoistIntoGlobals` runs before `SetEncoding`, so it doesn't get a chance to see the `set_encoding` op. Once I fix that, the next problem is that the `ConstExprAnalysis` just doesn't see any op as constant. My testcase is

```mlir
func.func @matmul_i8(%lhs: tensor<?x?xi8>, %rhs: tensor<?x?xi8>, %acc: tensor<?x?xi32>) -> tensor<?x?xi32> {
  %result = linalg.matmul ins(%lhs, %rhs: tensor<?x?xi8>, tensor<?x?xi8>) outs(%acc: tensor<?x?xi32>) -> tensor<?x?xi32>
  return %result: tensor<?x?xi32>
}
```

Patch to get HoistIntoGlobals to run after SetEncoding but before FormDispatchRegions:

```diff
--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp
@@ -266,7 +266,13 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
       .addPredicatedPass(clNormalizeInputIndexingMap,
                          createInterchangeTransposeGenericOpsPass)
       // Enable data tiling after all linalg level transformations.
-      .addPredicatedPass(clEnableDataTiling, createSetEncodingPass)
+      .addPredicatedPass(clEnableDataTiling, createSetEncodingPass);
+
+  if (transformOptions.constExprHoisting) {
+    passManager.addPass(IREE::Util::createHoistIntoGlobalsPass());
+  }
+
+  FunctionLikeNest(passManager)
```

Compile:

```
tools/iree-compile ~/matmul_i8.mlir -o /tmp/a.vmfb --iree-flow-enable-data-tiling --iree-hal-target-backends=llvm-cpu --iree-llvmcpu-enable-microkernels --mlir-print-ir-after-all --iree-opt-const-expr-hoisting --iree-opt-const-eval
```

Result - here is the IR that HoistIntoGlobal sees, but doesn't do anything about:

```mlir
  func.func @matmul_i8(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view, %arg2: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c0_i8 = arith.constant 0 : i8
    %c0_i32 = arith.constant 0 : i32
    %0 = hal.buffer_view.dim<%arg0 : !hal.buffer_view>[0] : index
    %1 = hal.buffer_view.dim<%arg0 : !hal.buffer_view>[1] : index
    %2 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<?x?xi8>{%0, %1}
    %3 = hal.buffer_view.dim<%arg1 : !hal.buffer_view>[0] : index
    %4 = hal.buffer_view.dim<%arg1 : !hal.buffer_view>[1] : index
    %5 = hal.tensor.import %arg1 "input 1" : !hal.buffer_view -> tensor<?x?xi8>{%3, %4}
    %6 = hal.buffer_view.dim<%arg2 : !hal.buffer_view>[0] : index
    %7 = hal.buffer_view.dim<%arg2 : !hal.buffer_view>[1] : index
    %8 = hal.tensor.import %arg2 "input 2" : !hal.buffer_view -> tensor<?x?xi32>{%6, %7}
    %9:2 = iree_linalg_ext.upper_bound_tile_size tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = LHS>> -> index, index
    %dim = tensor.dim %2, %c0 : tensor<?x?xi8>
    %10 = affine.apply #map()[%9#0, %dim]
    %dim_0 = tensor.dim %2, %c1 : tensor<?x?xi8>
    %11 = affine.apply #map()[%9#1, %dim_0]
    %padded = tensor.pad %2 low[0, 0] high[%10, %11] {
    ^bb0(%arg3: index, %arg4: index):
      tensor.yield %c0_i8 : i8
    } : tensor<?x?xi8> to tensor<?x?xi8>
    %12 = iree_linalg_ext.set_encoding %padded : tensor<?x?xi8> -> tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = LHS>>
```

Dumping the ConstExprAnalysis in HoistIntoGlobals.cpp, shows it's empty:

```
FOUND CONSTANTS:
----------------

```

**Comment by bjacob** [comment]
Note: I won't do further work here as I only have 1 week to go before my 3-week vacation and other things to finish. Hopefully someone else has cycles, this seems really important to me.

**Comment by bjacob** [comment]
Here's a proper testcase with a constant operand:

```mlir
func.func @matmul_i8(%lhs: tensor<100x100xi8>, %acc: tensor<100x100xi32>) -> tensor<100x100xi32> {
  %rhs = arith.constant dense<123> : tensor<100x100xi8>
  %result = linalg.matmul ins(%lhs, %rhs: tensor<100x100xi8>, tensor<100x100xi8>) outs(%acc: tensor<100x100xi32>) -> tensor<100x100xi32>
  return %result: tensor<100x100xi32>
}
```

WIth the above patch, this successfully hoists the `set_encoding` as a constant expression, and compiles down to `CleanupTensorShapes`, which complains about a `tensor.dim` that remains there, as, even though it is a constant expression, the `set_encoding` still has dynamic dims at this point, as they would only become static dims at `MaterializeEncoding`.

Note: as this is in the top-level function, at some point I wondered, maybe this only needs `MaterializeUpperBoundTileSize` --- but nope, this needs `MaterializeEncoding` because the shape here is the shape of the return value of the set_encoding that was hoisted out.

```
// -----// IR Dump After CleanupTensorShapes Failed (iree-flow-cleanup-tensor-shapes) //----- //
func.func @matmul_i8(%arg0: !hal.buffer_view, %arg1: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %c1 = arith.constant 1 : index
  %c0 = arith.constant 0 : index
  %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<100x100xi8>
  %1 = hal.tensor.import %arg1 "input 1" : !hal.buffer_view -> tensor<100x100xi32>
  %hoisted = util.global.load @hoisted : index
  %hoisted_0 = util.global.load @hoisted_0 : index
  %2 = affine.apply affine_map<()[s0] -> (s0 + 100)>()[%hoisted_0]
  %3 = affine.apply affine_map<()[s0] -> (s0 + 100)>()[%hoisted]
  %4 = flow.dispatch @matmul_i8_dispatch_0::@matmul_i8_dispatch_0_set_encoding_MATMUL_I8I8I32_LHS_DxD[%hoisted, %hoisted_0, %3, %2](%0, %hoisted, %hoisted_0, %3, %2) : (tensor<100x100xi8>, index, index, index, index) -> tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = LHS, original_type = tensor<100x100xi8>>>{%3, %2}
  %hoisted_3 = util.global.load @hoisted_3 : tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RHS, original_type = tensor<100x100xi8>>>
  %hoisted_4 = util.global.load @hoisted_4 : index
  %hoisted_5 = util.global.load @hoisted_5 : index
  %5 = affine.apply affine_map<()[s0] -> (s0 + 100)>()[%hoisted_4]
  %6 = affine.apply affine_map<()[s0] -> (s0 + 100)>()[%hoisted_5]
  %7 = flow.dispatch @matmul_i8_dispatch_1::@matmul_i8_dispatch_1_set_encoding_MATMUL_I8I8I32_RESULT_DxD[%hoisted_4, %hoisted_5, %5, %6](%1, %hoisted_4, %hoisted_5, %5, %6) : (tensor<100x100xi32>, index, index, index, index) -> tensor<?x?xi32, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RESULT, original_type = tensor<100x100xi32>>>{%5, %6}
  %dim = tensor.dim %hoisted_3, %c0 : tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RHS, original_type = tensor<100x100xi8>>>
  %dim_0 = tensor.dim %hoisted_3, %c1 : tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RHS, original_type = tensor<100x100xi8>>>
  %8 = flow.dispatch @matmul_i8_dispatch_2::@matmul_i8_dispatch_2_matmul_DxDxD_i8xi8xi32[%3, %2, %dim, %dim_0, %5, %6](%4, %hoisted_3, %7, %3, %2, %dim, %dim_0, %5, %6) : (tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = LHS, original_type = tensor<100x100xi8>>>{%3, %2}, tensor<?x?xi8, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RHS, original_type = tensor<100x100xi8>>>{%dim, %dim_0}, tensor<?x?xi32, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RESULT, original_type = tensor<100x100xi32>>>{%5, %6}, index, index, index, index, index, index) -> %7{%5, %6}
  %9 = flow.dispatch @matmul_i8_dispatch_3::@matmul_i8_dispatch_3_unset_encoding_MATMUL_I8I8I32_RESULT_DxD[%5, %6](%8, %5, %6) : (tensor<?x?xi32, #iree_linalg_ext.encoding<user = MATMUL_I8I8I32, role = RESULT, original_type = tensor<100x100xi32>>>{%5, %6}, index, index) -> tensor<100x100xi32>
  %10 = hal.tensor.export %9 "output 0" : tensor<100x100xi32> -> !hal.buffer_view
  return %10 : !hal.buffer_view
}

/usr/local/google/home/benoitjacob/matmul_i8_cst.mlir:3:13: error: 'tensor.dim' op unexpected during shape cleanup; dynamic dimensions must have been resolved prior to leaving the flow dialect
  %result = linalg.matmul ins(%lhs, %rhs: tensor<100x100xi8>, tensor<100x100xi8>) outs(%acc: tensor<100x100xi32>) -> tensor<100x100xi32>
```

**Comment by benvanik** [comment]
what's inserting those tensor.dims? could you post a full print-ir-after-all/change?
%hoisted_4/%hoisted_5 are the tensor dynamic dims, but if something inserts those tensor.dims _after_ they are materialized then things won't line up

**Comment by bjacob** [comment]
https://gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1

**Comment by bjacob** [comment]
I got an idea to unblock the actual users while we figure this out. Current users have dimensions that are static and are multiples of sufficiently large powers of two that they could use a flag of the form `--iree-flow-data-tiling-assume-all-tile-sizes-divide=16` that would allow SetEncoding to eschew padding.

**Comment by benvanik** [comment]
Ah, this is caused by the precise reason why I always had explicit captures on flow.dispatch.workgroups :) the flow.dispatch.region op doesn't capture dimensions and FormDispatchWorkgroups then needs to insert the dims but this happens _after_ the hoisting where the tensors have been made non-local.

Maybe we could swap the order of the passes around. Or we could add nasty tie_shape ops. Or we could go back to using explicit captures as intended as a way to avoid the need for such pass order dependencies or tie_shape-style metadata association ops.

**Comment by bjacob** [comment]
Is there a simple thing to try here instead of proceeding with my above idea of a short-term workaround flag or should I proceed ?

**Comment by benvanik** [comment]
(that is to say: for implicit captures to work there cannot be any hoisting/whole-program transformations that run prior to the pass that makes implicit -> explicit - here hoisting is running before that pass and that breaks the simplified assumptions of everything-in-a-single-function style code)

**Comment by benvanik** [comment]
We'll need to fix it properly (no flag) but I don't think we need to do that all in one step to start - I say go for the workaround and we make sure we're tracking removing it as part of the larger data tiling work.

**Comment by benvanik** [comment]
~~looking at the IR - flow.tensor.tie_shape inserted by the hoist globals pass may be the easiest solution - I'd hoped we'd gotten rid of the need for it outside of dispatches but so long as we have implicit flow.dispatch.region ops we'll probably require it. Yuck.~~ (nevermind, the hoist into globals pass doesn't know about dimensions yet - we'd probably have to insert it even earlier and preserve it much longer)

**Comment by MaheshRavishankar** [comment]
> Ah, this is caused by the precise reason why I always had explicit captures on flow.dispatch.workgroups :) the flow.dispatch.region op doesn't capture dimensions and FormDispatchWorkgroups then needs to insert the dims but this happens _after_ the hoisting where the tensors have been made non-local.
> 
> Maybe we could swap the order of the passes around. Or we could add nasty tie_shape ops. Or we could go back to using explicit captures as intended as a way to avoid the need for such pass order dependencies or tie_shape-style metadata association ops.

(still parsing this), but we cant swap this around :) . I am missing why this is a problem. The capture does happen and creates a `flow.dispatch.workgroups`. Maybe we just need to inline more operations. The `flow.dispatch.region` is just a way to break a mondo "create a dispatch and isolate from above" into two steps that helps. We use this to create a `flow.dispatch.workgroup` from this.

**Comment by MaheshRavishankar** [comment]
Could we discuss this on GVC sometime.. They might be a different solution than what is being discussed here.

**Comment by benvanik** [comment]
Yeah, let's chat. Issue is that the hoist pass runs and we end up losing the association of shape dimensions by the time the dispatch region formation happens.

For discussion:

After hoisting, this are dynamic and the association with the tensor dimensions are gone:
https://gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1#file-log-L1247-L1254

So when a dispatch region is formed and needs to get the dimensions it inserts tensor.dims, but now those can't resolve:
https://gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1#file-log-L1534-L1535

Running the CombineInitializers pass and a dedicated IPO pass for propagating shapes may help a bit as we could take the dims we have here and resolve the tensor.dims using them:
https://gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1#file-log-L1408

We could also propagate alignment (using util.align) across global stores/loads, if that helped.

**Comment by MaheshRavishankar** [comment]
Leaving some questions for chat.

> Yeah, let's chat. Issue is that the hoist pass runs and we end up losing the association of shape dimensions by the time the dispatch region formation happens.
> 
> For discussion:
> 
> After hoisting, this are dynamic and the association with the tensor dimensions are gone: [gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1#file-log-L1247-L1254](https://gist.github.com/bjacob/c54e460bc5cb624d58f95359316022a1#file-log-L1247-L1254)

Dont see how hoisting made this dynamic. It is dynamic because the shape depends on how the encoding is interpreted.

**Comment by hanhanW** [comment]
I'm going to work on this, and will start with some static cases. This might needs to revamp @bjacob 's https://github.com/openxla/iree/pull/14438 work, so we can set encoding on constants directly (w/o pad ops). I will try to plumb it through IREE, and then we can revisit how to make more progress on it.

**Comment by hanhanW** [comment]
I have a basic [prototype](https://github.com/openxla/iree/commit/210ed14c12efccb8f61a165b49776a64dcc39e8f) working locally:
  - Propagates original target devices to const_eval.
  - Disable const_eval on linalg_ext.upper_bound_tile_size ops

With #14438, I'm able to do const_eval for aligned cases.

A full IR dump can be found at https://gist.githubusercontent.com/hanhanW/d761ca3917e3c9ca96d383cc2f32aa08/raw

**Comment by benvanik** [comment]
Neat! We'll need to do proper attributes for encoding separate from the executable target info (and your getConstEvalExecutableTargetAttr and such) as we should be overriding specific behavior instead of doing big overrides. Basically, whatever is currently using ExecutableTargetAttr for encoding should instead be looking for an explicit specification and only falling back to the ExecutableTargetAttr for when it's not found.

**Comment by hanhanW** [comment]
> Basically, whatever is currently using ExecutableTargetAttr for encoding should instead be looking for an explicit specification and only falling back to the ExecutableTargetAttr for when it's not found.

This part is done in the prototype. The information is carried by an attribute (which is `iree.compiler.consteval.original.target`). It looks at `iree.compiler.consteval.original.target` to materialize encodings. If it is not found, it falls back to the `ExecutableTargetAttr`.

**Comment by benvanik** [comment]
That's different - you're overriding **all target information** - this should only be the information for encoding. Separation of concerns is important here as the target information is required for correctness and is not a hint.

**Comment by benvanik** [comment]
(think of it like parts of the code asking about encoding information should be looking precisely for encoding information - `iree_codegen.encoding.tile_size` or something - not looking up an ExecutableTargetAttr - having dozens+ places across the codebase looking up ExecutableTargetAttr effectively turns it into a global flag and that's the situation we want to avoid :)

**Comment by hanhanW** [comment]
> (think of it like parts of the code asking about encoding information should be looking precisely for encoding information - `iree_codegen.encoding.tile_size` or something - not looking up an ExecutableTargetAttr - having dozens+ places across the codebase looking up ExecutableTargetAttr effectively turns it into a global flag and that's the situation we want to avoid :)

Does it mean that we want to set `iree_codegen.encoding.tile_size` values in the [buildGlobalOptimizationPassPipeline](https://github.com/openxla/iree/blob/9e14c222a71610d2e8a246664996d4216ee9567b/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp#L152)? It would make `Flow/Transforms` depend on `Codegen/Common/CPU`, if we wanna query and set the tile_sizes to the attribute.

**Comment by hanhanW** [comment]
I have https://github.com/hanhanW/iree/commit/a646339dff3adeaffc34a9ec2743565195c20ece follow-up commit that introduces encoding_target in `ExecutableVariantOp`. It is an optional executable_target for materializing encodings.

The prototype depends on https://github.com/openxla/iree/pull/14438. I will send the rest of part for review after landing the parent PR.

**Comment by hanhanW** [comment]
The IR after all: https://gist.githubusercontent.com/hanhanW/b01f3b00b5911f1b1dcd415045b1c333/raw

With the `iree-flow-data-tiling-assume-tile-sizes-divisors-of` flag (in https://github.com/openxla/iree/pull/14438), we are able to generate set_encoding on constants directly:

```mlir
func.func @matmul_1024x512x256(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
  %cst = arith.constant 0.000000e+00 : f32
  %cst_0 = arith.constant dense<2.000000e+00> : tensor<256x512xf32>
  %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1024x256xf32>
  %1 = iree_linalg_ext.set_encoding %0 : tensor<1024x256xf32> -> tensor<1024x256xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS>>
  %2 = iree_linalg_ext.set_encoding %cst_0 : tensor<256x512xf32> -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
  %3 = tensor.empty() : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
  %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>) -> tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
  %5 = linalg.matmul ins(%1, %2 : tensor<1024x256xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS>>, tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>) outs(%4 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>) -> tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
  %6 = iree_linalg_ext.unset_encoding %5 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>> -> tensor<1024x512xf32>
  %7 = hal.tensor.export %6 "output 0" : tensor<1024x512xf32> -> !hal.buffer_view
  return %7 : !hal.buffer_view
}
```

After hoisting them to globals:

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu], iree.fixedpoint.iteration = 0 : index} {
  util.global private @hoisted : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
  func.func @matmul_1024x512x256(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1024x256xf32>
    %1 = iree_linalg_ext.set_encoding %0 : tensor<1024x256xf32> -> tensor<1024x256xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS>>
    %hoisted = util.global.load @hoisted : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    %2 = tensor.empty() : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
    %3 = linalg.fill ins(%cst : f32) outs(%2 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>) -> tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
    %4 = linalg.matmul ins(%1, %hoisted : tensor<1024x256xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS>>, tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>) outs(%3 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>) -> tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>>
    %5 = iree_linalg_ext.unset_encoding %4 : tensor<1024x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>> -> tensor<1024x512xf32>
    %6 = hal.tensor.export %5 "output 0" : tensor<1024x512xf32> -> !hal.buffer_view
    return %6 : !hal.buffer_view
  }
  util.initializer attributes {iree.compiler.consteval} {
    %cst = arith.constant dense<2.000000e+00> : tensor<256x512xf32>
    %0 = iree_linalg_ext.set_encoding %cst : tensor<256x512xf32> -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    util.global.store %0, @hoisted : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    util.initializer.return
  }
}
```

We set encoding_device to the `iree.compiler.consteval` initializers.

```mlir
  util.initializer attributes {iree.compiler.consteval, iree.compiler.consteval.encoding.target = #device_target_llvm_cpu} {
    %cst = arith.constant dense<2.000000e+00> : tensor<256x512xf32>
    %0 = iree_linalg_ext.set_encoding %cst : tensor<256x512xf32> -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    util.global.store %0, @hoisted : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    util.initializer.return
  }
```

We set `iree.compiler.consteval.encoding.target` attribute to the inner module op in JitGlobals, like what we've done for `hal.device.targets`.

In OutlineDispatchRegion pass, we propagates the attribute to `flow.executable` op.

```mlir
module attributes {hal.device.targets = [#hal.device.target<"vmvx", {executable_targets = [#hal.executable.target<"vmvx", "vmvx-bytecode-fb", {ukernels = false}>]}>], iree.compiler.consteval.encoding.target = #hal.device.target<"llvm-cpu", {executable_targets = [#hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>]}>, iree.consteval} {
  flow.executable private @jit_eval_dispatch_0 attributes {iree.compiler.consteval.encoding.target = #hal.device.target<"llvm-cpu", {executable_targets = [#hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>]}>} {
    flow.executable.export public @jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512 workgroups() -> (index, index, index) {
      %x, %y, %z = flow.dispatch.workgroup_count_from_slice 
      flow.return %x, %y, %z : index, index, index
    }
    builtin.module {
      func.func @jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512(%arg0: !flow.dispatch.tensor<readonly:tensor<256x512xf32>>, %arg1: !flow.dispatch.tensor<writeonly:tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>>) {
        %0 = flow.dispatch.tensor.load %arg0, offsets = [0, 0], sizes = [256, 512], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<256x512xf32>> -> tensor<256x512xf32>
        %1 = iree_linalg_ext.set_encoding %0 : tensor<256x512xf32> -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
        flow.dispatch.tensor.store %1, %arg1, offsets = [0, 0], sizes = [256, 512], strides = [1, 1] : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>> -> !flow.dispatch.tensor<writeonly:tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>>
        return
      }
    }
  }
  func.func @jit_eval(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<256x512xf32>
    %1 = flow.dispatch @jit_eval_dispatch_0::@jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512(%0) : (tensor<256x512xf32>) -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
    %2 = hal.tensor.export %1 "output 0" : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>> -> !hal.buffer_view
    return %2 : !hal.buffer_view
  }
}
```

Then we materialize them to `hal.executable` ops. The `hal.executable.variant` op optionally has `encoding_target` field.

```mlir
module attributes {hal.device.targets = [#hal.device.target<"vmvx", {executable_targets = [#hal.executable.target<"vmvx", "vmvx-bytecode-fb", {ukernels = false}>]}>], iree.compiler.consteval.encoding.target = #hal.device.target<"llvm-cpu", {executable_targets = [#hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>]}>, iree.consteval} {
  hal.executable private @jit_eval_dispatch_0 {
    hal.executable.variant public @vmvx_bytecode_fb, target = <"vmvx", "vmvx-bytecode-fb", {ukernels = false}>, encoding_target = <"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}> {
      hal.executable.export public @jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512 ordinal(0) layout(#hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer>]>]>) {
      ^bb0(%arg0: !hal.device):
        %x, %y, %z = flow.dispatch.workgroup_count_from_slice 
        hal.return %x, %y, %z : index, index, index
      }
      builtin.module {
        func.func @jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512() {
          %c0 = arith.constant 0 : index
          %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<256x512xf32>>
          %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>>
          %2 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [256, 512], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<256x512xf32>> -> tensor<256x512xf32>
          %3 = iree_linalg_ext.set_encoding %2 : tensor<256x512xf32> -> tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>
          flow.dispatch.tensor.store %3, %1, offsets = [0, 0], sizes = [256, 512], strides = [1, 1] : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>> -> !flow.dispatch.tensor<writeonly:tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>>>
          return
        }
      }
    }
  }
  func.func @jit_eval(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %c524288 = arith.constant 524288 : index
    %c0 = arith.constant 0 : index
    %c553648160_i32 = arith.constant 553648160 : i32
    %c1_i32 = arith.constant 1 : i32
    %c256 = arith.constant 256 : index
    %c512 = arith.constant 512 : index
    hal.buffer_view.assert<%arg0 : !hal.buffer_view> message("input 0") shape([%c256, %c512]) type(%c553648160_i32) encoding(%c1_i32)
    %0 = stream.tensor.import %arg0 : !hal.buffer_view -> tensor<256x512xf32> in !stream.resource<external>{%c524288}
    %1 = stream.resource.alloc uninitialized : !stream.resource<external>{%c524288}
    %2 = stream.cmd.execute with(%0 as %arg1: !stream.resource<external>{%c524288}, %1 as %arg2: !stream.resource<external>{%c524288}) {
      stream.cmd.dispatch @jit_eval_dispatch_0::@vmvx_bytecode_fb::@jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512 {
        ro %arg1[%c0 for %c524288] : !stream.resource<external>{%c524288},
        wo %arg2[%c0 for %c524288] : !stream.resource<external>{%c524288}
      } attributes {hal.interface.bindings = [#hal.interface.binding<0, 0>, #hal.interface.binding<0, 1>]}
    } => !stream.timepoint
    %3 = stream.timepoint.await %2 => %1 : !stream.resource<external>{%c524288}
    %4 = stream.tensor.export %3 : tensor<256x512xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>> in !stream.resource<external>{%c524288} -> !hal.buffer_view
    return %4 : !hal.buffer_view
  }
}
```

In CPUMaterializeEncoding pass, it looks at encoding_target firstly, if there are no encoding_target, the target will be used.

```mlir
func.func @jit_eval_dispatch_0_set_encoding_MATMUL_F32F32F32_RHS_256x512() {
  %c0 = arith.constant 0 : index
  %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<256x512xf32>>
  %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<32x256x16x1xf32>>
  %2 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [256, 512], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<256x512xf32>> -> tensor<256x512xf32>
  %3 = tensor.empty() : tensor<32x256x16x1xf32>
  %pack = tensor.pack %2 outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %3 : tensor<256x512xf32> -> tensor<32x256x16x1xf32>
  flow.dispatch.tensor.store %pack, %1, offsets = [0, 0, 0, 0], sizes = [32, 256, 16, 1], strides = [1, 1, 1, 1] : tensor<32x256x16x1xf32> -> !flow.dispatch.tensor<writeonly:tensor<32x256x16x1xf32>>
  return
}
```

**Comment by MaheshRavishankar** [comment]
Awesome @hanhanW ! Just following the flow and looking at IR after all this matches the touch points I had expected. Should wait for @benvanik to take a look, but most of the plumbing through Flow, etc. looks fine to me. Maybe only one thing to iterate on is the `encoding_target` attribute. It might come down to just a naming issue, but overall this looks about right.

**Comment by benvanik** [comment]
Getting there! May be worth a video chat as I still think the separation of concerns isn't quite connecting to the implementation. 

The HAL should not know about `encoding_target` and consteval also shouldn't have magic like `iree.compiler.consteval.encoding.target`. We really need to not be taking a whole device/executable target as input to the function that computes the number 16 and instead we need to distill out the minimum information required to make the decision and _only_ include that _only_ where it is required. The `hal.executable.variant` is only slightly better than a global compiler flag and including the complete configuration for what is a much more targeted bit of information is a big hammer that's very brittle (changes to what specific configurations hold/etc become load-bearing across a large swath of the codebase they shouldn't be, they don't generalize across backends as each backend can define its own configuration types, etc).

We should be starting from the bottom and asking questions like "in the pass that turns set_encoding into something else what is the exact bits of information it needs to do that?" and _add those bits to the set_encoding op_. It then gets much easier to work up the stack and find where that information is present and how to only plumb through what is required. A pass that ran and attributed each set_encoding op with exactly what it should be using (optional attr, etc) removes the need for big global flags and complex lookup logic across the codebase as by the time the set_encoding op is lowered all the information it needs is local.

It's important to get this right as when we go heterogenous we'll be evaluating single dispatches that may produce results for multiple other devices: we can't just slap a global flag and assume all set_encodings feeding into all produced values will have the same encoding. We also can't assume there's a single target that we can stamp on to consteval stuff as we may be evaluating constants for different targets in the same consteval pass and we really need them to deduplicate. If we have target A and B both using constant C we need to ensure we dedupe before that point by having a set_encoding op with the same encoding parameters derived from A and B CSE - if we can't do that because it carries the full target configuration _or_ we can't model that A and B may need different encodings everything breaks.

So concretely, instead of setting `iree.compiler.consteval.encoding.target` on `util.initializer` whatever is doing that should set the encoding parameters as tightly scoped as possible on the `set_encoding` ops themselves. Then nothing else changes anywhere and whatever is lowering `set_encoding` has what it needs. If a user wants to manually set the encoding parameters, or we want to perform mixed-device/mixed-target logic to derive generalized encodings, etc all that is just changes in whatever is setting the attributes on the ops and nothing involving the flow->stream->hal->codegen flow.

**Comment by benvanik** [comment]
(we'll need to solve the distillation of required information for extending this to other targets/deduping/etc, but an interim step would be to move your entire target configuration attr to a dialect attr on the set_encoding op - still keeps everything local and off of the intermediate ops but should work the same as what you have)

**Comment by MaheshRavishankar** [comment]
Well, we dont need the entire device encoding but there is a list of "cpu features" that are used to determine the size of the inner tile. We can start with that list of CPU features... I understand the concerns of heterogenous devices and scoping down the information as much as needed... Today it mostly looks for presence of certain CPU features. We can start with that maybe.

**Comment by benvanik** [comment]
We'll need to get one step further so that we can union/intersect and cover non-CPU targets - even if homogenous we can multi-target (and need to) x64 with AVX2 and AVX512 or all the various ARM configurations etc and don't want to either not pack anything because it's ambiguous or pack things multiple times. Getting the information more local to the op helps things but we can't roll this feature out any more broadly until we solve the join point here. I totally understand, though, that until we get _something_ we probably can't solve it and I think so long as we keep things local/prevent ossification throughout the codebase of what it currently is doing we can still make progress. Most important thing is that we keep sight on this being the real important problem to solve and that we keep edging closer to solving it even if we can't do it in one step!

**Comment by benvanik** [comment]
(I'm also not sure how to make any of this work outside of a demo with the target/CPU features, as the fact that llvm-cpu is doing the consteval _and_ we're only targeting llvm-cpu with the data tiling work is leading to some assumptions about what information is usable where - if we started evaluating constexprs with CUDA that backend isn't going to know how to deal with CPU features, or if we started targeting Vulkan/SPIR-V with data tiling the llvm-cpu backend isn't going to know what to do with SPIR-V feature flags)

**Comment by hanhanW** [comment]
(sorry that I forgot to put the other prototype commit to this thread: https://github.com/openxla/iree/commit/7560493411ecb1f286c6e1da8e8ffef0b78b887b)

That's why I keep `entire device` for encoding. For CPU side, cpu_features are currently used to determine "how we materialize the encodings". I believe that other backends will use similar things to determine it, but they are not "cpu features". So I make the `encoding_device` be `IREE::HAL::ExecutableTargetAttr` attribute, which gives us enough information, but not too general (to me).

RE heterogenous: the concerns are very valid to me! This is also why I strictly limit the number of targets to `1` and put a [TODO](https://github.com/openxla/iree/commit/7560493411ecb1f286c6e1da8e8ffef0b78b887b#diff-4ed89cc725e2e9106b6a9cf4f7c5016a0ef9608f949c44258050660429f95db8R40-R45) to `SetOriginalTargetOnGlobals` (maybe should rename it to something else). If they are just constants, we can evaluate them on whatever targets. The hard problem in this issue is that they have encodings. Different targets could materialize an encoding into different shapes/values. IMO, we should hoist them to globals and const-eval them for all the targets. Then the runtime side would have code like:

```
func.func @load_global_var():
  if target == 'avx2':
    %v = util.global.load @hoist_xxx_avx2  ...
    return %v
  if target == 'avx512:
    %v = util.global.load @hoist_xxx_avx512  ...
    return %v
  ...
```

**Comment by hanhanW** [comment]
(we can have a video chat for more details, I'm mostly free in the afternoon today and tomorrow.)

**Comment by benvanik** [comment]
Nah, needing effectively the entire set of compiler command line flags + target backend code to compute the encoding is definitely too general :)

We should be thinking of each component in the stack like this as defining a function that given some set of inputs produces some set of outputs (here "given this type, usage, and some traits of the target produce the encoding operations that such traits imply are good"). If you're in some random file in your codebase adding such a function your first instinct should not be to pass in all of the global state from everywhere in the program. I think we've skipped the step of thinking about what those inputs should be which is understandable given the complexity and sequencing of things but precisely what we need to solve as the rest is plumbing that is mostly in place now.

What we want to avoid is the example you give as it highlights one of the problems with this approach: we cannot be keying off of specific targets and instead need traits/properties/classes/categories/etc. Basically, if at some point something like the above switch is needed we've failed to solve the problem and prevented this from shipping by default. We know for a fact such a switch isn't required in other systems and shouldn't assume we'll need something like that - if anything we should be simpler! - so it's important to keep in mind that as we go lower down the stack if we find ourselves adding such hacks we're probably searching in the wrong solution space. Things should generally get simpler/more opaque as they lower as the cost for doing otherwise is the feature not working at all. I really just want to ensure we don't have to sit with this disabled for ages because of all the simplifying assumptions that don't hold outside of a very specific spike - we should be gearing up to use this on all backends and encouraging more extensions to it but as-is we definitely aren't ready to scale to that.

I'm free 3pm PST + today or any time tomorrow!

**Comment by hanhanW** [comment]
So we had an offline discussion. We may need an abstraction layer indicating the mapping between encoding tile_size/permutation and target instructions. So far we rely on cpu_features to drive it, but passing it to const_eval functions is too general. We probably need a new mechanism here.

The other solution is to materialize encodings before const-eval -- if there is only one executable target. I have a prototype: https://github.com/openxla/iree/commit/dec69c8512a2ef57feb93077f613ce57c239fa76

I added the new pass right after IPO pass.

The IR before `MaterializeHomogeneousEncodings`:

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
#map = affine_map<()[s0] -> ((1025 ceildiv s0) * s0 - 1025)>
#map1 = affine_map<()[s0] -> ((257 ceildiv s0) * s0 - 257)>
#map2 = affine_map<()[s0] -> ((513 ceildiv s0) * s0 - 513)>
#map3 = affine_map<()[s0] -> ((1025 ceildiv s0) * s0)>
#map4 = affine_map<()[s0] -> ((513 ceildiv s0) * s0)>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu], iree.fixedpoint.iteration = 0 : index} {
  func.func @matmul_1025x513x257(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant dense<2.000000e+00> : tensor<257x513xf32>
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1025x257xf32>
    %1:2 = iree_linalg_ext.upper_bound_tile_size tensor<1025x257xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS>> -> index, index
    %2 = affine.apply #map()[%1#0]
    %3 = affine.apply #map1()[%1#1]
    %padded = tensor.pad %0 low[0, 0] high[%2, %3] {
    ^bb0(%arg1: index, %arg2: index):
      tensor.yield %cst : f32
    } : tensor<1025x257xf32> to tensor<?x?xf32>
    %4 = iree_linalg_ext.set_encoding %padded : tensor<?x?xf32> -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<1025x257xf32>>>
    %5:2 = iree_linalg_ext.upper_bound_tile_size tensor<257x513xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS>> -> index, index
    %6 = affine.apply #map1()[%5#0]
    %7 = affine.apply #map2()[%5#1]
    %padded_1 = tensor.pad %cst_0 low[0, 0] high[%6, %7] {
    ^bb0(%arg1: index, %arg2: index):
      tensor.yield %cst : f32
    } : tensor<257x513xf32> to tensor<?x?xf32>
    %8 = iree_linalg_ext.set_encoding %padded_1 : tensor<?x?xf32> -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS, original_type = tensor<257x513xf32>>>
    %9:2 = iree_linalg_ext.upper_bound_tile_size tensor<1025x513xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT>> -> index, index
    %10 = affine.apply #map3()[%9#0]
    %11 = affine.apply #map4()[%9#1]
    %12 = tensor.empty(%10, %11) : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>>
    %13 = linalg.fill ins(%cst : f32) outs(%12 : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>>) -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>>
    %14 = linalg.matmul ins(%4, %8 : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = LHS, original_type = tensor<1025x257xf32>>>, tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RHS, original_type = tensor<257x513xf32>>>) outs(%13 : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>>) -> tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>>
    %15 = iree_linalg_ext.unset_encoding %14 : tensor<?x?xf32, #iree_linalg_ext.encoding<user = MATMUL_F32F32F32, role = RESULT, original_type = tensor<1025x513xf32>>> -> tensor<?x?xf32>
    %extracted_slice = tensor.extract_slice %15[0, 0] [1025, 513] [1, 1] : tensor<?x?xf32> to tensor<1025x513xf32>
    %16 = hal.tensor.export %extracted_slice "output 0" : tensor<1025x513xf32> -> !hal.buffer_view
    return %16 : !hal.buffer_view
  }
}
```

The IR after `MaterializeHomogeneousEncodings`:

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu], iree.fixedpoint.iteration = 0 : index} {
  func.func @matmul_1025x513x257(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant dense<2.000000e+00> : tensor<257x513xf32>
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1025x257xf32>
    %1 = tensor.empty() : tensor<65x257x16x1xf32>
    %pack = tensor.pack %0 padding_value(%cst : f32) inner_dims_pos = [0, 1] inner_tiles = [16, 1] into %1 : tensor<1025x257xf32> -> tensor<65x257x16x1xf32>
    %2 = tensor.empty() : tensor<33x257x16x1xf32>
    %pack_1 = tensor.pack %cst_0 padding_value(%cst : f32) outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %2 : tensor<257x513xf32> -> tensor<33x257x16x1xf32>
    %3 = tensor.empty() : tensor<65x33x16x16xf32>
    %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %5 = linalg.mmt4d ins(%pack, %pack_1 : tensor<65x257x16x1xf32>, tensor<33x257x16x1xf32>) outs(%4 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %6 = tensor.empty() : tensor<1025x513xf32>
    %unpack = tensor.unpack %5 inner_dims_pos = [0, 1] inner_tiles = [16, 16] into %6 : tensor<65x33x16x16xf32> -> tensor<1025x513xf32>
    %7 = hal.tensor.export %unpack "output 0" : tensor<1025x513xf32> -> !hal.buffer_view
    return %7 : !hal.buffer_view
  }
}
```

Then the pack ops on constant will be hoisted to globals.

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu], iree.fixedpoint.iteration = 0 : index} {
  util.global private @hoisted : tensor<33x257x16x1xf32>
  func.func @matmul_1025x513x257(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1025x257xf32>
    %1 = tensor.empty() : tensor<65x257x16x1xf32>
    %pack = tensor.pack %0 padding_value(%cst : f32) inner_dims_pos = [0, 1] inner_tiles = [16, 1] into %1 : tensor<1025x257xf32> -> tensor<65x257x16x1xf32>
    %hoisted = util.global.load @hoisted : tensor<33x257x16x1xf32>
    %2 = tensor.empty() : tensor<65x33x16x16xf32>
    %3 = linalg.fill ins(%cst : f32) outs(%2 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %4 = linalg.mmt4d ins(%pack, %hoisted : tensor<65x257x16x1xf32>, tensor<33x257x16x1xf32>) outs(%3 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %5 = tensor.empty() : tensor<1025x513xf32>
    %unpack = tensor.unpack %4 inner_dims_pos = [0, 1] inner_tiles = [16, 16] into %5 : tensor<65x33x16x16xf32> -> tensor<1025x513xf32>
    %6 = hal.tensor.export %unpack "output 0" : tensor<1025x513xf32> -> !hal.buffer_view
    return %6 : !hal.buffer_view
  }
  util.initializer attributes {iree.compiler.consteval} {
    %cst = arith.constant dense<2.000000e+00> : tensor<257x513xf32>
    %0 = tensor.empty() : tensor<33x257x16x1xf32>
    %cst_0 = arith.constant 0.000000e+00 : f32
    %pack = tensor.pack %cst padding_value(%cst_0 : f32) outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %0 : tensor<257x513xf32> -> tensor<33x257x16x1xf32>
    util.global.store %pack, @hoisted : tensor<33x257x16x1xf32>
    util.initializer.return
  }
}
```

There are two issues in the prototype.

## Dependencies

I added the pass to HAL/Transforms and it makes flow transforms "depends" on HAL transforms. Because the new pass will be inserted to `buildGlobalOptimizationPasses` section.

Based on the offline discussion, we want to split the flow pipeline into `FlowPreprocessing`, `buildGlobalOptimizationPass` and the rest of flow passes. I think we need to break it down to three phases and move `buildGlobalOptimizationPasses` to another directory. In this context, flow does not depend on HAL. The GlobalOptimizationPipiline depends on flow and HAL.

## Encoding is only supported on CPU

I hardcoded the backend to `llvm-cpu` in the prototype because it is only available on cpu backends. To generalize it, we need to add a `buildEncodingMaterializationPassPipeline` to HAL Target, and implement the method in each target backend. It is not easy in the prototype because of current code structure. We need to pass `TargetBackendRegistry` to the pass, but we don't have the information in `buildGlobalOptimizationPass`. We can do it properly after re-structure flow pipeline.


@benvanik is it okay to go with the prototype and refactor all the pipeline things later? Or the refactoring is a requirement for landing the prototype?

**Comment by MaheshRavishankar** [comment]
@hanhanW maybe we should also have a folder for that `pad -> pack` so that we get rid of the pad....

**Comment by hanhanW** [comment]
oops, I posted wrong IR. The new pass actually runs two passes. One is `CPUMaterializeUpperBoundTileSize` and the other is `CPUMaterializeEncoding`. The pad ops will be folded in the later pass, so we don't have to worry about it.

updated IR:

```mlir
#executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "cascadelake", cpu_features = "+cmov,+mmx,+popcnt,+sse,+sse2,+sse3,+ssse3,+sse4.1,+sse4.2,+avx,+avx2,+fma,+avx512f,+bmi,+bmi2,+aes,+pclmul,+avx512vl,+avx512bw,+avx512dq,+avx512cd,+avx512vnni,+adx,+clflushopt,+clwb,+cx16,+cx8,+crc32,+f16c,+fsgsbase,+fxsr,+invpcid,+lzcnt,+movbe,+pku,+prfchw,+rdrnd,+rdseed,+sahf,+x87,+xsave,+xsavec,+xsaveopt,+xsaves", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 64 : index, target_triple = "x86_64-unknown-unknown-eabi-elf", ukernels = true}>
#device_target_llvm_cpu = #hal.device.target<"llvm-cpu", {executable_targets = [#executable_target_embedded_elf_x86_64_]}>
module attributes {hal.device.targets = [#device_target_llvm_cpu], iree.fixedpoint.iteration = 0 : index} {
  func.func @matmul_1025x513x257(%arg0: !hal.buffer_view) -> !hal.buffer_view attributes {iree.abi.stub} {
    %cst = arith.constant 0.000000e+00 : f32
    %cst_0 = arith.constant dense<2.000000e+00> : tensor<257x513xf32>
    %0 = hal.tensor.import %arg0 "input 0" : !hal.buffer_view -> tensor<1025x257xf32>
    %1 = tensor.empty() : tensor<65x257x16x1xf32>
    %pack = tensor.pack %0 padding_value(%cst : f32) inner_dims_pos = [0, 1] inner_tiles = [16, 1] into %1 : tensor<1025x257xf32> -> tensor<65x257x16x1xf32>
    %2 = tensor.empty() : tensor<33x257x16x1xf32>
    %pack_1 = tensor.pack %cst_0 padding_value(%cst : f32) outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [16, 1] into %2 : tensor<257x513xf32> -> tensor<33x257x16x1xf32>
    %3 = tensor.empty() : tensor<65x33x16x16xf32>
    %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %5 = linalg.mmt4d ins(%pack, %pack_1 : tensor<65x257x16x1xf32>, tensor<33x257x16x1xf32>) outs(%4 : tensor<65x33x16x16xf32>) -> tensor<65x33x16x16xf32>
    %6 = tensor.empty() : tensor<1025x513xf32>
    %unpack = tensor.unpack %5 inner_dims_pos = [0, 1] inner_tiles = [16, 16] into %6 : tensor<65x33x16x16xf32> -> tensor<1025x513xf32>
    %7 = hal.tensor.export %unpack "output 0" : tensor<1025x513xf32> -> !hal.buffer_view
    return %7 : !hal.buffer_view
  }
}

```

**Comment by hanhanW** [comment]
There are many improvements, except one model in terms of artifact size!

https://github.com/openxla/iree/pull/14619#issuecomment-1675497753
